edgeless_node/src/agent/mod.rs

```diff
@@ -65,33 +67,104 @@ impl Agent {
         // value: provider_id
         let mut resource_instances = std::collections::HashMap::<edgeless_api::function_instance::ComponentId, String>::new();
 
+        // After spawning a new function, the function´s class is only used to determine which runner to start it on.
+        // When stopping, only the stop_function_id is provided which does not allow to know which runner it is
+        // currently deployed on. Here, we implement a instance_id -> function_class HashMap
+        let mut component_id_to_class_map = std::collections::HashMap::<edgeless_api::function_instance::ComponentId, String>::new();
+
         log::info!("Starting Edgeless Agent");
         while let Some(req) = receiver.next().await {
             match req {
                 AgentRequest::Spawn(spawn_req) => {
                     log::debug!("Agent Spawn {:?}", spawn_req);
-                    match runner.start(spawn_req).await {
-                        Ok(_) => {}
-                        Err(err) => {
-                            log::error!("Unhandled Start Error: {}", err);
+
+                    // Save function_class for further interaction.
+                    // We can assume that the Optional<instance_id> is present.
```

Comment:
The problem with `assert!()` in this part of the code is similar to the (wrongful) use of `break`, isn't it? As far as I understand the assertion, it makes the program panic effectively terminating the `main_task`. Thus, I'd recommend to make any AgentRequests fail gracefully, do you support this idea @ccicconetti ?

Reply:
The id will be always set here as it is set in the client part.
In general, everything we catch with an assert is a catastrophic failure.
We could move to assert_debug, but if you expect sth to happen do proper error handling.
https://doc.rust-lang.org/std/macro.assert.html

Reply:
General rule: assert!() should be used for invariants, that is to catch _software bugs_, hopefully in unit tests.

An assert that triggers in production is a catastrophe, but it would be even worse if there was no assert: an invariant has broken, this leads to unpredictable behavior and all kinds of problems. Crashing the application _may_ be marginally better than that.

I think assert_debug!() only makes sense if the check is expensive (not this case).

For this specific case, log::error!() and go on may be more appropriate than assert!().

My comment was more on the comment: it feels weird _for me_ to read "we assume x is true" in natural language rather than having assert!(x) ;-)

Reply:
As we cannot afford to panic the `main_task`, I added a sanitization clause in case the ID would not be existent. If this assumption can be trusted, I'd suggest removing the check as soon as `Optional<instance_id>` changes to `instance_id`.